packages/server-wallet/src/message-service/types.ts

```diff
@@ -0,0 +1,35 @@
+import {Message} from '@statechannels/client-api-schema';
+
+/**
+ * This is the handler that any messaging service implementation should call when receiving a message
+ * The handler is responsible for pushing the message into the appropriate wallet
+ */
+export type MessageHandler = (
+  // Including the to makes it easier for a messaging service to handle multiple wallets if it chooses to do so
+  to: string,
+  message: unknown,
+  messageService: MessageServiceInterface
```

Comment:
Why does `MessageHandler` require a message service? Do you plan to remove this in a future PR?

A `MessageService` should be able to:
1. allow the "owner" of the service to send messages to others
2. allow the "owner" to decide how to handle messages received

In production, each wallet has their own `MessageService` instance. Internally, it will listen for incoming messages. (eg. if it is on an HTTP server, it may listen to the `/inbox` endpoint. If it is an HTTP client, it may set up server sent events against a `/outbox` endpoint.) The `MessageHandler` allows the wallet to say "MessageService, when you receive a message, push it into my wallet."

Understandably, in our `with-peers` test suite, there needs to be some kind of a "broker" that has a handle on all of the `TestMessageService` implementations. In this PR, `TestMessageService` acts both as a broker _and_ the `MessageService` used by all of the peers. That has some undesirable side effects, like:
- if you `destroy()` the MessageService, it is destroyed for both peers

Reply:
> Why does `MessageHandler` require a message service? Do you plan to remove this in a future PR?

This is done to allow the `MessageHandler` to easily call `mesageService.send` if it wants to. Since the `MessageHandler` callback is constructed before the `MessageService` is constructed we don't have a reference to the `MessageService` in the `MessageHandler`. By passing it in it allow for easily constructing a `MessageHandler` that pushes messages into the wallet and calles `messageService.send` to respond.

> Understandably, in our with-peers test suite, there needs to be some kind of a "broker" that has a handle on all of the TestMessageService implementations. In this PR, TestMessageService acts both as a broker and the MessageService used by all of the peers. That has some undesirable side effects, like:

I think that's a fair point that is acting as both a broker and message service. It seemed the most straightforward to have the `TestMessageService` act as the broker instead of having a more complicated setup of two message services communicating.

> if you destroy() the MessageService, it is destroyed for both peers

I think that's not too big of a drawback? I think it's pretty apparent that destroying the message service would destroy it for both peers.

Reply:
You're declaring a `MessageHandler` type here that is not a test type. But, the type is suited to fit the `TestMessageService`, and not the production use case.

In production, a `MessageHandler`
1. will not have a handle on the peer's `MessageService`
2. does not need a handle on the "local `MessageService`

Right?

Reply:
> I think that's not too big of a drawback? I think it's pretty apparent that destroying the message service would destroy it for both peers.

Since the test setup shares a `MessageService`, yes, I agree it's apparent. But it's also not representational of the production use case, where each wallet has a distinct `MessageService` instance.

Reply:
> You're declaring a `MessageHandler` type here that is not a test type. But, the type is suited to fit the `TestMessageService`, and not the production use case.
> 
> In production, a `MessageHandler`
> 
> 1. will not have a handle on the peer's `MessageService`
> 2. does not need a handle on the "local `MessageService`
> 
> Right?

I think 2. is not true?

I'm under the assumption that a production message service might be setup roughly like this:

```
function handleMessage(message:Message){
     const response = this.wallet.pushMessage(message);
    await messageService.send(response);
}
const messageService = createMessageService(handleMessage);
```

However, to accomplish this the `handleMessage` needs a reference to the `messageService`. Since the function is being constructed before the message service is constructed, I've passed it in as an argument.

Reply:
I have a different (albeit fuzzy) model
```
// channel manager

function handleMessage(message:Message){
     return this.wallet.pushMessage(message);
}
const messageService = createMessageService(handleMessage);
wallet.on('outbox', message => messageService.send(message));
```

I think your approach is fine. Perhaps better, since the `wallet` does not have to emit events with your approach. What are the drawbacks of your approach?

In any case, the presence of the circular dependency indeed makes it confusing. Can you clarify that in the code?

Reply:
> I think your approach is fine. Perhaps better, since the `wallet` does not have to emit events with your approach. What are the drawbacks of your approach?

I think the main drawback is that it's kinda confusing?

  In our [discussion](https://www.notion.so/statechannels/State-Channels-a19d0ee8714e49e0b7b07e889e97cedd?p=ff9740bfa8f04d4981cf74774d944f33) we came to the conclusion that we wanted to avoid `EventEmitter` so I tried exploring solution without it. I think I still prefer avoiding the `EventEmitter` even with the confusion of the current solution.

> In any case, the presence of the circular dependency indeed makes it confusing. Can you clarify that in the code?

Yup! It also might be worthwhile writing a `readme.md` for the message service (especially since longer-term other people besides us might write a message service)

Reply:
An alternative is that the `MessageService` can be an event emitter that emits `'messageReceived'`. This allows the `ChannelManager` to attach a callback after the `MessageService` is instantiated.